### PR TITLE
[model, fsdp] fix: support Qwen3.5 linear attention under Ulysses SP

### PR DESCRIPTION
**Pull request by sunnweiwei**
### What does this PR do?

Current Qwen3.5/Qwen3.5-MoE training with FSDP remove-padding + Ulysses SP does not pass packed sequence boundaries into linear attention, so packed samples can share linear-attention state across sequence boundaries; the Qwen3.5 fused PPO loss can also re-slice already SP-sliced labels. This PR fixes those paths.

This PR:
- passes packed `cu_seqlens` into Qwen3.5 remove-padding forwards so linear attention does not cross packed sequence boundaries
- patches Qwen3.5 decoder/GatedDeltaNet forwards to use FLA packed kernels and CP context under Ulysses SP
- avoids re-slicing pre-sliced `shift_labels` in Qwen3.5 fused PPO loss
- adds a small distributed regression test for packed Qwen3.5 linear attention under Ulysses SP, including the torch conv1d fallback path

### Checklist Before Starting

- [x] Search for similar PRs. Query link: https://github.com/verl-project/verl/pulls?q=is%3Apr+qwen3_5+ulysses+sp
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Test

Environment used for Qwen3.5 validation: 8x H100, FLA 0.5.1 at commit `c525f49`, transfer-queue 0.1.7, transformers 5.3.0.dev0.

- `python3 -m compileall -q verl/models/transformers/qwen3_5.py verl/models/transformers/monkey_patch.py verl/workers/engine/fsdp/transformer_impl.py tests/special_distributed/test_qwen35_linear_attention_ulysses_sp.py` -> passed
- `python3 -m pytest -q tests/models/test_fused_kernels_ulysses_sp_on_cpu.py` -> 21 passed
- `torchrun --nproc_per_node=8 -m pytest -svv tests/special_distributed/test_qwen35_linear_attention_ulysses_sp.py` -> 2 passed, covering both `causal_conv1d_fn` and `torch_conv1d_fallback`
  - causal conv1d path: sequence-cut output err ratio `3.61e-4`, grad err ratio `8.39e-4`; many-short output err ratio `5.58e-4`, grad err ratio `1.30e-3`
  - torch conv1d fallback: sequence-cut output err ratio `3.76e-4`, grad err ratio `9.05e-4`; many-short output err ratio `5.44e-4`, grad err ratio `1.319e-3`
- 8x H100 Qwen3.5-27B GRPO sanity run: batch size 64, rollout `n=8`, response length 8192, PPO mini-batch 16, `ppo_max_token_len_per_gpu=16384`, vLLM TP=4, gpu memory util 0.75, raw rollout logprobs/bypass. SP8 completed without OOM; reward step2-9 mean `0.948`, step time `215.8s`, throughput `595 tokens/s`.

### API and Usage Example

No public API changes. Existing FSDP + Ulysses SP configs can be used with Qwen3.5, for example:

```bash
actor_rollout_ref.actor.use_remove_padding=True actor_rollout_ref.actor.use_fused_kernels=True actor_rollout_ref.actor.fused_kernel_options.impl_backend=torch actor_rollout_ref.actor.ulysses_sequence_parallel_size=8
```

### Design & Code Changes

- `verl/workers/engine/fsdp/transformer_impl.py`: pass packed `cu_seqlens`/`cu_seqlens_cpu` for Qwen3.5/Qwen3.5-MoE remove-padding forwards, including padded total length under SP.
- `verl/models/transformers/qwen3_5.py`: add packed Qwen3.5 linear-attention forward support for FLA packed/CP kernels, preserve packed sequence boundaries for fallback paths, and avoid double slicing `shift_labels`.
- `verl/models/transformers/monkey_patch.py`: install the Qwen3.5 decoder and GatedDeltaNet forward patches.
- `tests/special_distributed/test_qwen35_linear_attention_ulysses_sp.py`: compare full forward vs Ulysses SP forward on boundary-aligned, sequence-cut, single-sequence, and many-short packed cases, and check output/input-gradient error ratios.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): not run because the validation Docker image does not include `pre_commit`/`ruff`; targeted `compileall` and pytest checks are listed above.
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs). No docs update needed; this fixes existing Qwen3.5 model support and does not add a user-facing API.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. Added a GPU distributed regression test under `tests/special_distributed`; it requires `torchrun` + GPUs + FLA, so it may need to run in the special distributed/GPU test environment rather than CPU CI.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). Not requested from this environment.
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`. Not recipe-related.

### CODE REVIEW

File: verl/models/transformers/qwen3_5.py [hunk 1]
```diff
@@ -33,6 +38,328 @@
...
+            torch.full((end - start,), i, device=cu_seqlens.device, dtype=torch.int32)
+            for i, (start, end) in enumerate(zip(offsets[:-1], offsets[1:], strict=True))
+        ]
+        return torch.cat(seq_idx, dim=0).unsqueeze(0)
+
+
+def _split_packed_args(cu_seqlens: torch.LongTensor, tensors: tuple[torch.Tensor, ...]):
+    offsets = cu_seqlens.detach().cpu().tolist()
+    for start, end in zip(offsets[:-1], offsets[1:], strict=True):
+        yield tuple(tensor[:, start:end] for tensor in tensors)
```
[the remark this answers is not in the document]
Comment (sunnweiwei) [reply verl/models/transformers/qwen3_5.py]: Fixed in `bac87e7f`. `_split_packed_args` now accepts both `dim` and `cu_seqlens_cpu`; the conv fallback passes `dim=2`, and the distributed regression test now covers the torch conv1d fallback path explicitly.

File: verl/models/transformers/qwen3_5.py [hunk 2]
```diff
@@ -33,6 +38,328 @@
...
+        return seq_idx
+    prefix_idx = seq_idx[:, :1].expand(seq_idx.shape[0], prefix_len)
+    return torch.cat((prefix_idx, seq_idx), dim=1)
+
+
+def _packed_causal_conv1d_fallback(self, mixed_qkv: torch.Tensor, cu_seqlens: torch.LongTensor):
+    outputs = []
+    for (segment,) in _split_packed_args(cu_seqlens, (mixed_qkv,)):
+        outputs.append(F.silu(self.conv1d(segment)[:, :, : segment.shape[-1]]))
+    return torch.cat(outputs, dim=-1)
```
[the remark this answers is not in the document]
Comment (sunnweiwei) [reply verl/models/transformers/qwen3_5.py]: Fixed in `bac87e7f`. `_packed_causal_conv1d_fallback` now slices packed segments along sequence dim 2 and receives `cu_seqlens_cpu` to avoid repeated GPU-to-CPU offset copies.

File: verl/models/transformers/qwen3_5.py [hunk 3]
```diff
@@ -33,6 +38,328 @@
 logger.setLevel(os.getenv("VERL_LOGGING_LEVEL", "WARN"))
 
 
+def _call_accepts_kwarg(fn, name: str) -> bool:
+    try:
+        params = signature(fn).parameters
+    except (TypeError, ValueError):
+        return True
+    return name in params or any(param.kind == param.VAR_KEYWORD for param in params.values())
```
[the remark this answers is not in the document]
Comment (sunnweiwei) [reply verl/models/transformers/qwen3_5.py:43]: Fixed in `bac87e7f`. Removed the defensive `try`/`except` around `signature(fn)`.

File: verl/models/transformers/qwen3_5.py [hunk 4]
```diff
@@ -33,6 +38,328 @@
...
+        return self.chunk_gated_delta_rule(query, key, value, **kwargs)
+
+    outputs = []
+    for q_i, k_i, v_i, g_i, beta_i in _split_packed_args(cu_seqlens, (query, key, value, g, beta)):
+        split_kwargs = dict(kwargs)
+        split_kwargs["g"] = g_i
+        split_kwargs["beta"] = beta_i
+        out_i, _ = self.chunk_gated_delta_rule(q_i, k_i, v_i, **split_kwargs)
+        outputs.append(out_i)
+    return torch.cat(outputs, dim=1), None
```
[the remark this answers is not in the document]
Comment (sunnweiwei) [reply verl/models/transformers/qwen3_5.py:190]: Fixed in `bac87e7f`. The chunk fallback now passes `cu_seqlens_cpu` into `_split_packed_args`.

File: verl/models/transformers/qwen3_5.py [hunk 5]
```diff
@@ -33,6 +38,328 @@
...
+            )
+            mixed_qkv = conv_output[..., conv_prefix_len:] if conv_prefix_len else conv_output
+        elif model_cu_seqlens is not None:
+            if cp_context is not None:
+                mixed_qkv, conv_prefix_len = _prepend_cp_conv_prefix(mixed_qkv, cp_context)
+                model_cu_seqlens = model_cu_seqlens + conv_prefix_len
+                model_cu_seqlens[0] = 0
+                mixed_qkv = _packed_causal_conv1d_fallback(self, mixed_qkv, model_cu_seqlens)[..., conv_prefix_len:]
+            else:
+                mixed_qkv = _packed_causal_conv1d_fallback(self, mixed_qkv, model_cu_seqlens)
```
[the remark this answers is not in the document]
Comment (sunnweiwei) [reply verl/models/transformers/qwen3_5.py]: Fixed in `bac87e7f`. The CP conv prefix fallback now keeps `model_cu_seqlens_cpu` in sync with `model_cu_seqlens` and passes it through to `_packed_causal_conv1d_fallback`.

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 6]
```diff
@@ -937,6 +937,8 @@ def prepare_model_inputs(self, micro_batch: TensorDict):
         multi_modal_inputs = extract_multi_modal_inputs(micro_batch.get("multi_modal_inputs", []))
         input_ids = micro_batch["input_ids"]
         position_ids = micro_batch["position_ids"]
+        hf_model_type = getattr(self.model_config.hf_config, "model_type", None)
+        pass_packed_cu_seqlens = hf_model_type in {"qwen3_5", "qwen3_5_moe"}
```
Comment (Luosuu) [inline verl/workers/engine/fsdp/transformer_impl.py]: why we have to have special handling for qwen3_5 cu_seqlens?

Comment (wuxibin89) [reply verl/workers/engine/fsdp/transformer_impl.py]: `fsdp/transformer_impl.py` should be model agnostic.

**Review by wuxibin89 — CHANGES_REQUESTED** [review]

File: verl/workers/engine/fsdp/transformer_impl.py (hunk 6 above)
Comment (sunnweiwei) [reply verl/workers/engine/fsdp/transformer_impl.py]: Agreed, fixed in 3f1ecb21b4c83efad32a6d945dd9b7776743d245. `fsdp/transformer_impl.py` no longer checks Qwen3.5 model types. It now passes `cu_seqlens` only when the model forward explicitly declares that argument, so the engine stays model agnostic.

**Comment by yueyiming2009** [comment]
@sunnweiwei @wuxibin89 We've been waiting for this PR to be merged to fix #6780. But since it has taken a long time, shall we send a separate targeted fix for #6780?

**Comment by sunnweiwei** [comment]
> fsdp/transformer_impl.py should be model agnostic.  @wuxibin89 

fixed in 3f1ecb21b4c83efad32a6d945dd9b7776743d245

**Review by wuxibin89 — DISMISSED** [review]

**Review by wuxibin89 — APPROVED** [review]
